### ISSUE TITLE
Question about the nnclr.py

### ISSUE THREAD

**Issue by HuangChiEn** [vturrisi/solo-learn#252]
Thanks for making this SSL benchmark public, it's amazing!!

I just want to know about the detail of the nnclr.py implementation.
Why do we need to define the queue as the static class variable in pl.LightningModule ?

**Comment by DonkeyShot21** [comment]
Hi! Thank you for using the library.

In our implementation it is not a static class variable, it is a buffer registered in the method class. AFAIK this is the best method to ensure:
1) gradients are not propagated to the queue and it is not added to the optimizer even by mistake
2) the queue is saved in the checkpoint
3) the queue is on the right device

**Comment by HuangChiEn** [comment]
> 
Thanks for your reply.


Actually, I try to implement the nnclr method alone. However, the nnclr can not converge (only got 2% Top1 ACC). When I compare with the nnclr.py in solo-learn, I found that I forgot to declare the queue as the static variable at the begin of class (then wrap it into the self.regist_buffer).

Although the performance is still not good (maybe caused by pretrain on cifar100), but it can achieve about "20% Top-1" ACC in pre-training on cifar100 with 100 epochs and batch size 256.

So,  i wonder why this (declare as static variable) help so much
```
class NNCLR(BaseMethod):
    # define as static var here with type annotation..
    queue: torch.Tensor

    def __init__(
        self,
        proj_output_dim: int,
        proj_hidden_dim: int,
        pred_hidden_dim: int,
        temperature: float,
        queue_size: int,
        **kwargs
    ):
        super().__init__(**kwargs)

        self.temperature = temperature
        self.queue_size = queue_size

        # projector
        self.projector = nn.Sequential(
            nn.Linear(self.features_dim, proj_hidden_dim),
            nn.BatchNorm1d(proj_hidden_dim),
            nn.ReLU(),
            nn.Linear(proj_hidden_dim, proj_hidden_dim),
            nn.BatchNorm1d(proj_hidden_dim),
            nn.ReLU(),
            nn.Linear(proj_hidden_dim, proj_output_dim),
            nn.BatchNorm1d(proj_output_dim),
        )

        # predictor
        self.predictor = nn.Sequential(
            nn.Linear(proj_output_dim, pred_hidden_dim),
            nn.BatchNorm1d(pred_hidden_dim),
            nn.ReLU(),
            nn.Linear(pred_hidden_dim, proj_output_dim),
        )

        # queue, also in here overwrite the self.queue with regist-buffer
        self.register_buffer("queue", torch.randn(self.queue_size, proj_output_dim))
        self.register_buffer("queue_y", -torch.ones(self.queue_size, dtype=torch.long))
        # normlize
        self.queue = F.normalize(self.queue, dim=1)
        self.register_buffer("queue_ptr", torch.zeros(1, dtype=torch.long))
``` 

It's cool that remove the static declaration drop the accuracy of the model..

**Comment by HuangChiEn** [comment]
> Hi! Thank you for using the library.
> 
> In our implementation it is not a static class variable, it is a buffer registered in the method class. AFAIK this is the best method to ensure:
> 
> 1. gradients are not propagated to the queue and it is not added to the optimizer even by mistake
> 2. the queue is saved in the checkpoint
> 3. the queue is on the right device

Thanks for your reply~
